src/libraries/System.Private.CoreLib/src/System/IO/FileStatus.Unix.cs

```diff
@@ -10,44 +10,46 @@ internal partial struct FileStatus
+        private const int InitializedExists = -2;    // entry exists.
+        private const int InitializedNotExists = -1; // no entry exists.
+        private const int Uninitialized = 0;         // uninitialized, '0' to make default(FileStatus) uninitialized.
+
+        // Tracks the initialization state.
+        // < 0 : initialized succesfully. Value is InitializedExists or InitializedNotExists.
+        //   0 : uninitialized.
+        // > 0 : initialized with error. Value is raw errno.
+        private int _initialized;
+
+        // The last cached lstat information about the file.
+        // Must only be used after calling EnsureCachesInitialized and checking FileExists is true.
         private Interop.Sys.FileStatus _fileCache;
 
-        // -1: if the file cache isn't initialized - Refresh should always change this value
-        //  0: if the file cache was initialized with no errors
-        // Positive number: the error code returned by the lstat call
-        private int _initializedFileCache;
-
-        // We track intent of creation to know whether or not we want to (1) create a
-        // DirectoryInfo around this status struct or (2) actually are part of a DirectoryInfo.
-        // Set to true during initialization when the DirectoryEntry's INodeType describes a directory
-        internal bool InitiallyDirectory { get; set; }
-
-        // Is a directory as of the last refresh
-        // Its value can come from either the main path or the symbolic link path
+        // Is a directory as of the last refresh.
+        // Its value can come from either the main path or the symbolic link target.
+        // Must only be used after calling EnsureCachesInitialized.
         private bool _isDirectory;
 
-        // Exists as of the last refresh
-        private bool _exists;
+        private bool FileExists => _initialized == InitializedExists;
 
-        private bool IsFileCacheInitialized => _initializedFileCache == 0;
+        // Check if the main path (without following symlinks) has the hidden attribute set.
+        private bool HasHiddenFlag
+        {
+            get
+            {
+                Debug.Assert(_initialized != Uninitialized); // Use this after EnsureCachesInitialized has been called.
 
-        // Check if the main path (without following symlinks) has the hidden attribute set
-        // Ideally, use this if Refresh has been successfully called at least once.
-        // But since it is also used for soft retrieval during FileSystemEntry initialization,
-        // we return false early if the cache has not been initialized
-        internal bool HasHiddenFlag => IsFileCacheInitialized &&
-            (_fileCache.UserFlags & (uint)Interop.Sys.UserFlags.UF_HIDDEN) == (uint)Interop.Sys.UserFlags.UF_HIDDEN;
+                return FileExists && (_fileCache.UserFlags & (uint)Interop.Sys.UserFlags.UF_HIDDEN) == (uint)Interop.Sys.UserFlags.UF_HIDDEN;
```

Comment:
Previously, this and some other methods used `IsFileCacheInitialized`, though they should have checked `_exists`.